azure/services/aso/interfaces.go

```diff
@@ -25,15 +25,15 @@ import (
 )
 
 // Reconciler is a generic interface used to perform reconciliation of Azure resources backed by ASO.
-type Reconciler interface {
-	CreateOrUpdateResource(ctx context.Context, spec azure.ASOResourceSpecGetter, serviceName string) (result genruntime.MetaObject, err error)
-	DeleteResource(ctx context.Context, spec azure.ASOResourceSpecGetter, serviceName string) (err error)
+type Reconciler[T genruntime.MetaObject] interface {
+	CreateOrUpdateResource(ctx context.Context, spec azure.ASOResourceSpecGetter[T], serviceName string) (result T, err error)
+	DeleteResource(ctx context.Context, spec azure.ASOResourceSpecGetter[T], serviceName string) (err error)
 }
 
 // TagsGetterSetter represents an object that supports tags.
-type TagsGetterSetter interface {
+type TagsGetterSetter[T genruntime.MetaObject] interface {
 	GetAdditionalTags() infrav1.Tags
-	GetDesiredTags(resource genruntime.MetaObject) (infrav1.Tags, error)
-	GetActualTags(resource genruntime.MetaObject) (infrav1.Tags, error)
-	SetTags(resource genruntime.MetaObject, tags infrav1.Tags) error
+	GetDesiredTags(resource T) infrav1.Tags
+	GetActualTags(resource T) infrav1.Tags
+	SetTags(resource T, tags infrav1.Tags)
```

Comment:
This is one of the biggest wins for generics here. Since the only way these methods could have errored before was due to unexpected types, those would now be compiler errors so we don't have to do any error handling at runtime to catch those same errors.